### PR TITLE
feat: add SelfCleaningRegistry and make Registry interface type safe

### PR DESCRIPTION
**Pull request by pk-work**
This change introduces a SelfCleaningRegistry. The SelfCleaningRegistry knows which values were added by which cluster node. When a node leaves the cluster, all node-related values are removed from the registry.

This change also changes the Registry interface. Now a Registry is type safe and offers supports registration of multiple values for the same key.

### CODE REVIEW

**Review by halber — CHANGES_REQUESTED** [review]

File: src/main/java/io/neonbee/entity/EntityVerticle.java [hunk 1]
```diff
@@ -254,15 +247,35 @@ public void start(Promise<Void> promise) {
...
+        return announceOrConcealEntities(entityType -> getRegistry(vertx).unregister(entityType, getQualifiedName()));
+    }
+
+    @Override
+    public void stop(Promise<Void> stopPromise) throws Exception {
+        concealEntityVerticle().compose(v -> Future.<Void>future(stopPromiseFromSuper -> {
+            try {
+                super.stop(stopPromiseFromSuper);
+            } catch (Exception e) {
+                stopPromise.fail(e);
```
[the remark this answers is not in the document]
Comment (pk-work) [reply src/main/java/io/neonbee/entity/EntityVerticle.java:267]: Good catch

File: src/main/java/io/neonbee/NeonBee.java [hunk 2]
```diff
@@ -593,11 +597,7 @@ private Future<Void> deployModules() {
 
         this.healthRegistry = new HealthCheckRegistry(vertx);
         this.modelManager = new EntityModelManager(this);
-        if (vertx.isClustered()) {
-            this.entityRegistry = new ClusterEntityRegistry(vertx, EntityVerticle.REGISTRY_NAME);
-        } else {
-            this.entityRegistry = new WriteSafeRegistry<>(vertx, EntityVerticle.REGISTRY_NAME);
-        }
+        this.entityRegistry = new SelfCleaningRegistry<>(vertx, EntityVerticle.REGISTRY_NAME);
```
[the remark this answers is not in the document]
Comment (pk-work) [reply src/main/java/io/neonbee/NeonBee.java:600]: We can do this, but then we also have to pass the HealthCheckRegistry into the NeonBee constructor.

File: src/main/java/io/neonbee/internal/registry/SelfCleaningRegistryController.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+    @SuppressWarnings("PMD.AvoidSynchronizedAtMethodLevel")
+    private static synchronized SelfCleaningRegistryController getInstance(Vertx vertx) {
+        // the second condition is required for testing
+        if (instance == null || !Objects.equals(vertx, instance.vertx)) {
+            instance = new SelfCleaningRegistryController(vertx);
+        }
+        return instance;
+    }
```
[the remark this answers is not in the document]
Comment (pk-work) [reply src/main/java/io/neonbee/internal/registry/SelfCleaningRegistryController.java:196]: We have to consider here whether we want to have a singleton or multiple instances (one per registry instance) of SelfCleaningRegistryController.

Both is fine and easy to implement. Let's have a quick poll tomorrow in the Committer Meeting.

**Review by halber — APPROVED** [review]
LGTM

**Pull request merged**